shared/src/ledger/protocol/transactions/votes/update.rs

```diff
@@ -0,0 +1,547 @@
+    pub fn new(
+        votes: Votes,
+        voting_powers: &HashMap<(Address, BlockHeight), FractionalVotingPower>,
+    ) -> Result<Self> {
+        let mut inner = HashMap::default();
+        for (address, block_height) in votes {
+            let fract_voting_power = match voting_powers
+                .get(&(address.clone(), block_height))
+            {
+                Some(fract_voting_power) => fract_voting_power,
+                None => {
+                    return Err(eyre!(
+                        "No fractional voting power provided for vote by \
+                         validator {address} at block height {block_height}"
+                    ));
+                }
+            };
+            _ = inner
+                .insert(address, (block_height, fract_voting_power.to_owned()));
+        }
+        Ok(Self { inner })
+    }
+
+    pub fn voters(&self) -> BTreeSet<Address> {
+        self.inner.keys().cloned().collect()
+    }
+
+    /// Consumes `self` and returns a `VoteInfo` with any addresses from
+    /// `voters` removed, as well as the set of addresses that were actually
+    /// removed. Useful for removing voters who have already voted for
+    /// something.
+    pub fn without_voters<'a>(
+        self,
+        voters: impl IntoIterator<Item = &'a Address>,
+    ) -> (Self, HashSet<&'a Address>) {
+        let mut inner = self.inner;
+        let mut removed = HashSet::default();
+        for voter in voters {
+            inner.remove(voter);
+            removed.insert(voter);
+        }
+        (Self { inner }, removed)
+    }
+}
+
+impl IntoIterator for NewVotes {
+    type IntoIter = std::collections::hash_set::IntoIter<Self::Item>;
+    type Item = (Address, BlockHeight, FractionalVotingPower);
+
+    fn into_iter(self) -> Self::IntoIter {
+        let items: HashSet<_> = self
+            .inner
+            .into_iter()
+            .map(|(address, (block_height, fract_voting_power))| {
+                (address, block_height, fract_voting_power)
+            })
+            .collect();
+        items.into_iter()
+    }
+}
+
+/// Calculate an updated [`Tally`] based on one that is in storage under `keys`,
+/// with new votes from `vote_info` applied, as well as the storage keys that
+/// would change. If [`Tally`] is already `seen = true` in storage, then no
+/// votes from `vote_info` should be applied, and the returned changed keys will
+/// be empty.
+pub(in super::super) fn calculate<D, H, T>(
+    store: &mut Storage<D, H>,
+    keys: &vote_tallies::Keys<T>,
+    vote_info: NewVotes,
+) -> Result<(Tally, ChangedKeys)>
+where
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+    T: BorshDeserialize,
+{
+    tracing::info!(
+        ?keys.prefix,
+        validators = ?vote_info.voters(),
+        "Recording validators as having voted for this event"
+    );
+    let tally_pre = super::storage::read(store, keys)?;
+    if tally_pre.seen {
```

Comment:
could we change `seen` to something like `approved`? and `seen_by` to something along the lines of `voted_by`. this is a bit more generic. I've noticed `seen` and `seen_by` in the logs for validator set updates, and thought it was a bit silly, since it makes more sense in the context of ethereum events, rather than valset upds

Reply:
:100:% agree, the term "seen" is overloaded in our design/code.